src/oop/special_class_method/hard/q4/MainTest.java

```diff
@@ -0,0 +1,326 @@
+        assertTrue(testClass.hasField(string2FieldName, String.class),
+                "Your " + string2FieldName + " does not have the correct type.");
+        assertTrue(testClass.hasModifier(string2FieldName, "private"),
+                "Your " + string2FieldName + " does not have the correct visibility modifier.");
+        assertTrue(testClass.hasField(stringFieldName),
+                "The attribute " +  stringFieldName + " could not be found in your class. Please make sure you have added it, and it is spelled correctly.");
+        assertTrue(testClass.hasField(stringFieldName, String.class),
+                "Your " + stringFieldName + " does not have the correct type.");
+        assertTrue(testClass.hasModifier(stringFieldName, "private"),
+                "Your " + stringFieldName + " does not have the correct visibility modifier.");
+    }
+
+    private static Stream<Arguments> constructorInputProvider() {
+        return Stream.of(
+                Arguments.of("Test String", "Long"),
+                Arguments.of("MyString", "Round"),
+                Arguments.of("First", "Second"),
+                Arguments.of("Third", "Fourth")
+
+        );
+    }
+
+    @Test
+    public void detergentHasRequiredConstructor() {
+        String missingConstructorMessage = "The %s class is missing a required constructor. Make sure that it is named correctly and has the correct parameters.";
+        String wrongAccessModifier = "The %s class constructor has the wrong visibility modifier. Make sure that it is visible from all other classes.";
+        Class<?>[] constructorArgs = new Class[]{String.class, String.class};
+        assertTrue(testClass.hasConstructor(constructorArgs), String.format(missingConstructorMessage, className));
+        assertTrue(testClass.hasModifier(constructorArgs, "public"), String.format(wrongAccessModifier, className));
+    }
+
+    @ParameterizedTest
+    @MethodSource("constructorInputProvider")
+    @Tag("dependent1")
+    public void detergentClassHasCorrectConstructor(String value1, String value2) throws Throwable {
+        Object[][] arguments = {
+                {value1, String.class},
+                {value2, String.class}
+        };
+        Object classInstance = testClass.createInstance(arguments);
+        String incorrectString1ValueMessage =
+                "Your " + className + " constructor does not initialize the " + string2FieldName + " attribute to the correct value.";
+        String incorrectString2ValueMessage =
+                "Your " + className + " constructor does not initialize the " + stringFieldName + " attribute to the correct value.";
+        _assertEquals(value1, testClass.getFieldValue(classInstance, string2FieldName),
+                incorrectString1ValueMessage);
+        _assertEquals(value2, testClass.getFieldValue(classInstance, stringFieldName),
+                incorrectString2ValueMessage);
+    }
+
+    @Test
```

Comment:
everything tested here is already tested in detergentClassHasCorrectSetNameMethod, so there is no need for this test